### DISCUSSION TITLE
The possible way for periodic refreshing.

### DISCUSSION THREAD

**Discussion by JimCSuen** [eclipse-glsp/glsp#1461] [Q&A]
Sorry for asking another question. 

I am struggling to realize a periodic updating logic on the source model file so that the rendering could be real-time updated (no matter according to the source model file or by querying other (e.g., online) resources). I find that most operations or service logic is one-time, temporary, and passively triggered by user activities. I need help with where, which kind of contribution/handler/action/operation, or how, could I use them, to write these kinds of automatic, periodic, and active re-render behavior logic in this delicate framework.

**Comment by martin-fleck-at** [comment]
Hi @JimCSuen!

So basically what you need is to reload the model from "source" (wherever that is) and send it again to the client, is that right? If you need to trigger this behavior from the client, you need to simply send a `RequestModelAction` for the same source URI. We use that, for instance, in the [GLSP diagram widget for Theia to offer a reload behavior](https://github.com/eclipse-glsp/glsp-theia-integration/blob/8ac9d6a8325fd6f67b1aa3f9d7dbf773dcb4c0a0/packages/theia-integration/src/browser/diagram/glsp-diagram-widget.ts#L277). That action is handled on the server and basically, [re-reads the content from the source model storage and submits it to the client](https://github.com/eclipse-glsp/glsp-server-node/blob/90fd7589ce82d2a5b7b0a57696baba5830a3ad26/packages/server/src/common/features/model/request-model-action-handler.ts#L49) again. If you want to trigger the same behavior from the server side, you can either do the same manually and submit the model through the model submission handler or write a dummy operation since the [default operation action handler](https://github.com/eclipse-glsp/glsp-server-node/blob/90fd7589ce82d2a5b7b0a57696baba5830a3ad26/packages/server/src/common/operations/operation-action-handler.ts#L66-L72) does basically that. In general if you want to send a new model from the server to the client you should look into the `ModelSubmissionHandler`.

I hope this helps you with your periodic refresh behavior.

**Reply by JimCSuen** [reply]
Hi @martin-fleck-at! Thanks for answering me :)

Now I know that to call `reload` if I want to re-load the source model storage. So, it is reloaded from the source-model-storage instead of the source file where the model is stored, right? Then, even if the source file is changed, it has nothing to do with the source-model-storage?

**Reply by martin-fleck-at** [reply]
That is correct, in GLSP we use the [SourceModelStorage](https://github.com/eclipse-glsp/glsp-server-node/blob/90fd7589ce82d2a5b7b0a57696baba5830a3ad26/packages/server/src/common/features/model/source-model-storage.ts#L34) as an abstraction on how to load the model. Our [default implementation](https://github.com/eclipse-glsp/glsp-server-node/blob/90fd7589ce82d2a5b7b0a57696baba5830a3ad26/packages/server/src/node/gmodel/gmodel-storage.ts#L31) simply interprets the source uri given as part of the `RequestModelAction` as a file URI and tries to load the model from there. However, if you were to interpret the options given be the action differently, you may load your model from another external source, e.g., a database. It really is just a thin load/save persistence layer that by default reads a file but could be anything.

So if you call a reload, and we reload the model from the source model storage and the source model storage re-reads the file, you would get the latest version of the file, independent from where it was changed. But the source model storage is not responsible for detecting those changes.

**Reply by JimCSuen** [reply]
Now I see. I will try this with a (maybe, dummy) action that contains a periodic reload operation within a dead loop. I will give my feedback as soon as I have any progress :)
